"Cannot convert node to SQL Comprehension" error when calling .length after groupBy + map

Reproduced in Slick  3.1:

``` scala
//Single field table/entity
case class TestInstance(id: String)

class TestQuery(tag: Tag) extends Table[TestInstance](tag, "test_query") {
  def id = column[String]("id")

  override def * = (id) <> (TestInstance.apply, TestInstance.unapply)
}

object TestQuery {
  val table = TableQuery[TestQuery]
  val q = table.groupBy(_.id).map { case (c, tbl)  =>
    (c, tbl.length)
  }

  q.result //this runs just fine producing following query: "select `id`, count(1) from `test_query` group by `id`"
  q.length.result //<-- Here we'll get an "Cannot convert node to SQL Comprehension" error.
}
```

But if I change problem line with the following:

``` scala
q.drop(0).length.result
```

Everything works fine and query returns correct result (count of grouped rows).

Full Stack Trace of the error:

```
[10.11.2015 15:27:21.088] [ERROR] [application-akka.actor.default-dispatcher-12] [application] Error 'internalError - Cannot convert node to SQL Comprehension
| GroupBy t9 : Vector[(t9<String'>, Vector[t17<{s18: String'}>])]
|   from s8: Bind : Vector[t17<{s18: String'}>]
|     from s13: Table test_query : Vector[@t11<{id: String'}>]
|     select: Pure t17 : Vector[t17<{s18: String'}>]
|       value: StructNode : {s18: String'}
|         s18: Path s13.id : String'
|   by: Path s8.s18 : String'
' occured.
slick.SlickTreeException: Cannot convert node to SQL Comprehension
| GroupBy t9 : Vector[(t9<String'>, Vector[t17<{s18: String'}>])]
|   from s8: Bind : Vector[t17<{s18: String'}>]
|     from s13: Table test_query : Vector[@t11<{id: String'}>]
|     select: Pure t17 : Vector[t17<{s18: String'}>]
|       value: StructNode : {s18: String'}
|         s18: Path s13.id : String'
|   by: Path s8.s18 : String'

    at slick.compiler.MergeToComprehensions$$anonfun$26.apply(MergeToComprehensions.scala:167) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$26.apply(MergeToComprehensions.scala:167) ~[slick_2.11-3.1.0.jar:na]
    at scala.Option.getOrElse(Option.scala:121) ~[scala-library-2.11.7.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$convertBase$1(MergeToComprehensions.scala:167) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeFilterWhere$1$2.apply(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeFilterWhere$1$2.apply(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeFilterWhere$1(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeGroupBy$1(MergeToComprehensions.scala:156) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeSortBy$1$10.apply(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeSortBy$1$10.apply(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeSortBy$1(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$10.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$10.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeTakeDrop$1(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$createTopLevel$1(MergeToComprehensions.scala:253) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$createSourceOrTopLevel$1$1.apply(MergeToComprehensions.scala:236) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$createSourceOrTopLevel$1$1.apply(MergeToComprehensions.scala:234) ~[slick_2.11-3.1.0.jar:na]
    at scala.Option.getOrElse(Option.scala:121) ~[scala-library-2.11.7.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$createSourceOrTopLevel$1(MergeToComprehensions.scala:234) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$convertBase$1(MergeToComprehensions.scala:166) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeFilterWhere$1$2.apply(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeFilterWhere$1$2.apply(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeFilterWhere$1(MergeToComprehensions.scala:161) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeGroupBy$1(MergeToComprehensions.scala:146) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeSortBy$1$10.apply(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeSortBy$1$10.apply(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeSortBy$1(MergeToComprehensions.scala:96) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$10.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$10.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:344) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeTakeDrop$1(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$9.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$slick$compiler$MergeToComprehensions$$mergeTakeDrop$1$9.apply(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.mergeCommon(MergeToComprehensions.scala:319) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$mergeTakeDrop$1(MergeToComprehensions.scala:68) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$createTopLevel$1(MergeToComprehensions.scala:253) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.slick$compiler$MergeToComprehensions$$convert1$1(MergeToComprehensions.scala:269) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.convert(MergeToComprehensions.scala:289) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$apply$1$$anonfun$15.apply(MergeToComprehensions.scala:29) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$apply$1$$anonfun$15.apply(MergeToComprehensions.scala:28) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.ClientSideOp$.mapResultSetMapping(ClientSideOp.scala:24) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.ClientSideOp$$anonfun$mapResultSetMapping$1.apply(ClientSideOp.scala:25) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.ClientSideOp$$anonfun$mapResultSetMapping$1.apply(ClientSideOp.scala:25) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.UnaryNode$class.mapChildren(Node.scala:224) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.First.mapChildren(ClientSideOp.scala:31) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.First.nodeMapServerSide(ClientSideOp.scala:35) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.First.nodeMapServerSide(ClientSideOp.scala:31) ~[slick_2.11-3.1.0.jar:na]
    at slick.ast.ClientSideOp$.mapResultSetMapping(ClientSideOp.scala:25) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$apply$1.apply(MergeToComprehensions.scala:28) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions$$anonfun$apply$1.apply(MergeToComprehensions.scala:28) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.CompilerState.map(QueryCompiler.scala:226) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.MergeToComprehensions.apply(MergeToComprehensions.scala:28) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler$$anonfun$runPhase$1.apply(QueryCompiler.scala:79) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler$$anonfun$runPhase$1.apply(QueryCompiler.scala:78) ~[slick_2.11-3.1.0.jar:na]
    at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58) ~[scala-library-2.11.7.jar:na]
    at slick.ast.SymbolNamer.use(Symbol.scala:112) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler.runPhase(QueryCompiler.scala:78) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler$$anonfun$3.apply(QueryCompiler.scala:65) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler$$anonfun$3.apply(QueryCompiler.scala:63) ~[slick_2.11-3.1.0.jar:na]
    at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:155) ~[scala-library-2.11.7.jar:na]
    at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:155) ~[scala-library-2.11.7.jar:na]
    at scala.collection.Iterator$class.foreach(Iterator.scala:742) ~[scala-library-2.11.7.jar:na]
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1194) ~[scala-library-2.11.7.jar:na]
    at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:155) ~[scala-library-2.11.7.jar:na]
    at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1194) ~[scala-library-2.11.7.jar:na]
    at slick.compiler.QueryCompiler.runPhases(QueryCompiler.scala:63) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler.run(QueryCompiler.scala:53) ~[slick_2.11-3.1.0.jar:na]
    at slick.compiler.QueryCompiler.run(QueryCompiler.scala:47) ~[slick_2.11-3.1.0.jar:na]
    at slick.profile.BasicProfile$API$class.repQueryActionExtensionMethods(BasicProfile.scala:52) ~[slick_2.11-3.1.0.jar:na]
    at slick.driver.JdbcProfile$$anon$1.repQueryActionExtensionMethods(JdbcProfile.scala:69) ~[slick_2.11-3.1.0.jar:na]
```

--- a/slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/AggregateTest.scala
+++ b/slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/AggregateTest.scala
@@ -48,6 +48,9 @@ class AggregateTest extends AsyncTest[RelationalTestDB] {
       val q1 = q0.map(_._2.length).sortBy(identity)
       db.run(mark("q1", q1.result)).map { r0t: Seq[Int] => r0t shouldBe Vector(2, 3, 3) }
     }.flatMap { _ =>
+      val q0 = ts.groupBy(_.a).map(_._1).length
+      db.run(mark("q0", q0.result)).map { r0t: Int => r0t shouldBe 3 }
+    }.flatMap { _ =>
       val q = (for {
         (k, v) <- ts.groupBy(t => t.a)
       } yield (k, v.length, v.map(_.a).sum, v.map(_.b).sum)).sortBy(_._1)

--- a/slick/src/main/scala/slick/compiler/CreateAggregates.scala
+++ b/slick/src/main/scala/slick/compiler/CreateAggregates.scala
@@ -62,7 +62,7 @@ class CreateAggregates extends Phase {
 
   /** Recursively inline mapping Bind calls under an Aggregate */
   def inlineMap(a: Aggregate): Aggregate = a.from match {
-    case Bind(s1, f1, Pure(StructNode(defs1), ts1)) =>
+    case Bind(s1, f1, Pure(StructNode(defs1), ts1)) if !f1.isInstanceOf[GroupBy] => // mergeToComprehensions always needs a Bind around a GroupBy
       logger.debug("Inlining mapping Bind under Aggregate", a)
       val defs1M = defs1.iterator.toMap
       val sel = a.select.replace({

--- a/slick/src/main/scala/slick/compiler/ForceOuterBinds.scala
+++ b/slick/src/main/scala/slick/compiler/ForceOuterBinds.scala
@@ -50,7 +50,7 @@ class ForceOuterBinds extends Phase {
   }
 
   def maybewrap(n: Node): Node = n match {
-    case _: Join | _: Pure | _: Union | _: FilteredQuery | _:TableNode => wrap(n)
+    case _: Join | _: Pure | _: Union | _: FilteredQuery | _:TableNode | _: GroupBy => wrap(n)
     case _ => nowrap(n)
   }
 }

--- a/slick/src/main/scala/slick/compiler/MergeToComprehensions.scala
+++ b/slick/src/main/scala/slick/compiler/MergeToComprehensions.scala
@@ -150,8 +150,13 @@ class MergeToComprehensions extends Phase {
         (c2, replacements)
 
       case n @ Pure(Aggregate(s1, f1, str1), ts) =>
-        val (c1, replacements1) = mergeFilterWhere(f1, true)
         logger.debug("Merging Aggregate source into Comprehension:", Ellipsis(n, List(0, 0)))
+        val (c1, replacements1) = {
+          val (c, r) = mergeFilterWhere(f1, true)
+          if(c.groupBy.isDefined || c.distinct.isDefined || c.fetch.isDefined || c.offset.isDefined) toSubquery(c, r)
+          else (c, r)
+        }
+        logger.debug("New source at this point is:", c1)
         val str2 = applyReplacements(str1, replacements1, c1)
         val c2 = c1.copy(select = Pure(str2, ts)).infer()
         logger.debug("Merged Aggregate source into Comprehension:", c2)

